### ISSUE TITLE
[Windows] crash reports in KeyboardKeyEmbedderHandler::KeyboardHookImpl

### ISSUE THREAD

**Issue by tgucio** [flutter/flutter#108731]
This is now coming in fairly large numbers from users via Microsoft App Store:

`null_class_ptr_read_c0000005_flutter_windows.dll!flutter::keyboardkeyembedderhandler::keyboardhookimpl`

```
Frame	Image	Function	Offset
0	flutter_windows.dll	flutter::KeyboardKeyEmbedderHandler::KeyboardHookImpl	0x000000000000006A	
1	flutter_windows.dll	flutter::KeyboardKeyEmbedderHandler::KeyboardHook	0x0000000000000093	
2	flutter_windows.dll	flutter::KeyboardKeyHandler::KeyboardHook	0x000000000000017C	
3	flutter_windows.dll	flutter::FlutterWindowsView::SendKey	0x000000000000011A	
4	flutter_windows.dll	flutter::FlutterWindowsView::OnKey	0x000000000000009F	
5	flutter_windows.dll	flutter::FlutterWindowWin32::OnKey	0x000000000000008F	
6	flutter_windows.dll	flutter::KeyboardManagerWin32::PerformProcessEvent	0x000000000000010B	
7	flutter_windows.dll	flutter::KeyboardManagerWin32::ProcessNextEvent	0x000000000000008D	
8	flutter_windows.dll	std::_Func_impl_no_alloc_`lambda at ../../flutter/shell/platform/windows/keyboard_manager_win32.cc:366:7',void,bool_::_Do_call	0x000000000000003E	
9	flutter_windows.dll	std::_Func_impl_no_alloc_`lambda at ../../flutter/shell/platform/windows/flutter_windows_view.cc:457:7',void,bool_::_Do_call	0x0000000000000069	
10	flutter_windows.dll	flutter::KeyboardKeyHandler::ResolvePendingEvent	0x00000000000000FA	
11	flutter_windows.dll	std::_Func_impl_no_alloc_`lambda at ../../flutter/shell/platform/windows/keyboard_key_channel_handler.cc:146:25',void,const unsigned char *,unsigned long long_::_Do_call	0x000000000000007A	
12	flutter_windows.dll	flutter::BinaryMessengerImpl::Send::_lambda_0_::__invoke	0x000000000000003A	
13	flutter_windows.dll	std::_Func_impl_no_alloc_fml::internal::CopyableLambda_`lambda at ../../flutter/shell/platform/embedder/embedder_platform_message_response.cc:27:25'_,void_::_Do_call	0x0000000000000053	
14	flutter_windows.dll	flutter::EmbedderTaskRunner::PostTask	0x0000000000000114	
15	flutter_windows.dll	FlutterEngineRunTask	0x000000000000000F	
16	flutter_windows.dll	std::_Func_impl_no_alloc_`lambda at ../../flutter/shell/platform/windows/flutter_windows_engine.cc:160:37',void,const FlutterTask *_::_Do_call	0x000000000000001B	
17	flutter_windows.dll	flutter::TaskRunner::ProcessTasks	0x0000000000000103	
18	flutter_windows.dll	flutter::TaskRunnerWin32::ProcessTasks	0x0000000000000014	
19	flutter_windows.dll	flutter::TaskRunnerWin32Window::ProcessTasks	0x0000000000000076	
20	flutter_windows.dll	flutter::TaskRunnerWin32Window::HandleMessage	0x0000000000000015	
21	user32.dll	UserCallWinProcCheckWow	0x00000000000002F8	
22	user32.dll	DispatchMessageWorker	0x0000000000000249	
```

No known way to reproduce yet.

**Comment by tgucio** [comment]
Close similitudes with https://github.com/flutter/flutter/issues/97789 although the crash location in `KeyboardHookImpl ` offset is slightly different. Plus, we get some line numbers from the symbol files submitted to the store.

**Comment by danagbemava-nc** [comment]
Hi @tgucio, what version of flutter was used to deploy the version of the app you're experiencing this issue with?

**Comment by cbracken** [comment]
@dkwingsmt assigned your way -- are you able to take a look at this?

**Comment by tgucio** [comment]
Another batch of reports coming as "heap corruption":

```
heap_corruption_c0000374_flutter_windows.dll!_free_base

Frame	Image	Function	Offset
0	ntdll.dll	RtlReportFatalFailure	0x0000000000000009
1	ntdll.dll	RtlReportCriticalFailure	0x0000000000000097
2	ntdll.dll	RtlpHeapHandleError	0x0000000000000012
3	ntdll.dll	RtlpHpHeapHandleError	0x000000000000007A
4	ntdll.dll	RtlpLogHeapFailure	0x0000000000000045
5	ntdll.dll	RtlpHpLfhSubsegmentFreeBlock	0x00000000000800C7
6	ntdll.dll	RtlpFreeHeapInternal	0x00000000000003D6
7	ntdll.dll	RtlpHpFreeWithExceptionProtection	0x000000000000001E
8	ntdll.dll	RtlFreeHeap	0x000000000000006D
9	flutter_windows.dll	_free_base	0x000000000000001C
10	flutter_windows.dll	std::_Tree_std::_Tmap_traits_unsigned long long,unsigned long long,std::less_unsigned long long_,std::allocator_std::pair_const unsigned long long,unsigned long long_ _,0_ _::erase	0x0000000000000049
11	flutter_windows.dll	flutter::KeyboardKeyEmbedderHandler::KeyboardHookImpl	0x000000000000027D
12	flutter_windows.dll	flutter::KeyboardKeyEmbedderHandler::KeyboardHook	0x0000000000000093
13	flutter_windows.dll	flutter::KeyboardKeyHandler::KeyboardHook	0x000000000000019A
14	flutter_windows.dll	flutter::FlutterWindowsView::SendKey	0x000000000000011A
15	flutter_windows.dll	flutter::FlutterWindowsView::OnKey	0x000000000000009F
16	flutter_windows.dll	flutter::FlutterWindow::OnKey	0x000000000000008F
17	flutter_windows.dll	flutter::KeyboardManager::PerformProcessEvent	0x000000000000010B
18	flutter_windows.dll	flutter::KeyboardManager::ProcessNextEvent	0x000000000000008D
19	flutter_windows.dll	flutter::KeyboardManager::HandleMessage	0x00000000000005AD
20	flutter_windows.dll	flutter::Window::HandleMessage	0x0000000000000077
21	user32.dll	UserCallWinProcCheckWow	0x000000000000036C
22	user32.dll	DispatchMessageWorker	0x00000000000001F1
```

**Comment by cbracken** [comment]
@tgucio do you have a reduced repro for this, or repro steps we could use to figure out what's going on? This looks like we're maybe hitting a double free.

**Comment by yaakovschectman** [comment]
I can reproduce a possibly identical issue by starting up the Flutter app on Windows, then opening up Windows Narrator (CTRL+WIN+ENTER), then repeatedly and quickly hitting CAPS once it is open.

**Comment by tgucio** [comment]
@cbracken So far the closes I got to mini repro was https://github.com/flutter/flutter/issues/97789 although that occasionally required a fair amount of monkey testing on the keyboard.

I'll check @yaakovschectman's suggestion that CapsLock may be the key key here.

**Comment by yaakovschectman** [comment]
What I think is happening is that when the screen reader sends a simulated caps lock keypress to the app, our `get_key_state_` method still reports that the key with virtual keycode `0x14` (caps lock) is not pressed, so `SynchronizeCritialPressedStates` removes its entry from `pressingRecords_` ( https://github.com/flutter/engine/blob/0738285e9f9f90aa235cf4e59c661b3a615b3425/shell/platform/windows/keyboard_key_embedder_handler.cc#L460 ) before the `assert` we are looking at, which causes it to fail.
The invariant we seem to expect that an entry in `pressingRecords_` is present from the keydown event until the corresponding keyup event is violated by that `SynchronizeCritialPressedStates` when Narrator sends a caps lock keystroke. The solution might be to only `assert(record_iter != pressingRecords_.end());` when the virtual key is not in `critical_keys_`, and otherwise just do a runtime check and skip erasing if it's already erased.

**Comment by yaakovschectman** [comment]
Addressed by [flutter/engine#36129](https://github.com/flutter/engine/pull/36129), patching the only known case where this issue can be reproduced. I will mark as closed, as this resolves that case.
